altos-rust/altos-core/libs/heap/free_list_allocator/src/free_list.rs

```diff
@@ -265,12 +269,70 @@ impl FreeList {
                 },
             }
         });
-        match previous {
-            Some(previous) => {
-                previous.next_block = alloc_block;
-                alloc_block.get_ref_mut().unwrap().next_block = Link::from(current);
+
+
+        // Determine if the deallocated block is adjacent to the leading free block.
+        let merge_with_previous = match previous {
+            Some(ref previous) => {
+              previous.as_ptr() as usize + previous.block_size == alloc_block.as_ptr() as usize
             },
-            None => {
+            None => false
+        };
+
+        // Determine if the deallocated block is adjadent to the following(tail) free block.
+        let merge_with_current = match current {
+            Some(ref current) => {
+                alloc_block.as_ptr() as usize + alloc_block.get_ref().unwrap().block_size
```

Comment:
Instead of `alloc_block.get_ref().unwrap().block_size` we could just use `used_memory`. Seems a lot more concise and gets rid of that `unwrap()` call which might imply that we're doing a potentially dangerous unwrap of an optional (which we're not since we're creating the block up above.)